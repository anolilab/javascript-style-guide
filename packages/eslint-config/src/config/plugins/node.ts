import type { OptionsFiles, OptionsOverrides, OptionsPackageJson } from "../../types";
import { createConfig } from "../../utils/create-config";
import interopDefault from "../../utils/interop-default";

// @see https://github.com/eslint-community/eslint-plugin-n
export default createConfig<OptionsFiles & OptionsOverrides & OptionsPackageJson>("all", async (config, oFiles) => {
    const { files = oFiles, overrides, packageJson } = config;

    const pluginNode = await interopDefault(import("eslint-plugin-n"));

    const nodeVersion = packageJson?.engines?.["node"];

    return [
        {
            name: "anolilab/node/setup",
            plugins: {
                n: pluginNode,
            },
        },
        {
            files,
            name: "anolilab/node/rules",
            rules: {
                ...pluginNode.configs.recommended.rules,

                // https://eslint.org/docs/rules/global-require
                "global-require": "error",

                // enforce return after a callback
                "n/callback-return": "off",

                // enforce the style of file extensions in import declarations
                // This rule is buggy @see https://github.com/eslint-community/eslint-plugin-n/issues/21
                "n/file-extension-in-import": "off",

                // enforces error handling in callbacks (node environment)
                "n/handle-callback-err": "off",

                // Redundant with `import/no-extraneous-dependencies`.
                "n/no-extraneous-import": "off",

                // disallow require() expressions which import extraneous modules
                "n/no-extraneous-require": "off",

                // require all requires be top-level
                // Redundant with `import/no-unresolved`.
                "n/no-missing-import": "off", // This rule is also buggy and doesn't support `node:`.

                // disallow require() expressions which import non-existence modules
                "n/no-missing-require": "off",

                // disallow use of the Buffer() constructor
                // disallow mixing regular variable and require declarations
                "n/no-mixed-requires": [
                    "error",
                    {
                        allowCall: true,
                        grouping: true,
                    },
                ],

                // disallow use of new operator with the require function
                "n/no-new-require": "error",

                // disallow use of process.env
                "n/no-process-env": "off",

                // disallow string concatenation with __dirname and __filename
                // unicorn/no-process-exit is enabled instead.
                "n/no-process-exit": "off",

                // restrict usage of specified node modules
                "n/no-restricted-modules": "off",

                // disallow process.exit()
                // disallow use of synchronous methods (off by default)
                "n/no-sync": "off",

                // disallow bin files that npm ignores
                "n/no-unpublished-bin": "error",

                // require that process.exit() expressions use the same code path as throw
                "n/process-exit-as-throw": "error",

                // https://eslint.org/docs/rules/no-buffer-constructor
                "no-buffer-constructor": "error",

                // https://eslint.org/docs/rules/no-path-concat
                "no-path-concat": "error",

                ...nodeVersion
                    ? {
                        "n/no-unsupported-features/es-builtins": ["error", { version: nodeVersion }],
                        "n/no-unsupported-features/es-syntax": ["error", { ignores: ["modules"], version: nodeVersion }],
                        "n/no-unsupported-features/node-builtins": ["error", { version: nodeVersion }],
                    }
                    : {},

                ...overrides,
            },
        },
    ];
});
