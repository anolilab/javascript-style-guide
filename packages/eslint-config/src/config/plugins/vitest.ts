import type {
    OptionsFiles,
    OptionsHasPrettier,
    OptionsIsInEditor,
    OptionsOverrides,
    OptionsTypeScriptWithTypes,
} from "../../types";
import { createConfig } from "../../utils/create-config";
import interopDefault from "../../utils/interop-default";
import vitestGlobals from "../../utils/vitest-globals";

// Hold the reference so we don't redeclare the plugin on each call
// eslint-disable-next-line @typescript-eslint/no-explicit-any
let pluginTest: any;

export default createConfig<OptionsFiles & OptionsHasPrettier & OptionsIsInEditor & OptionsOverrides & OptionsTypeScriptWithTypes>(
    "vitest",
    async (config, oFiles) => {
        const {
            files = oFiles,
            isInEditor = false,
            overrides,
            prettier,
            tsconfigPath,
        } = config;

        const [vitestPlugin, noOnlyTestsPlugin] = await Promise.all([
            interopDefault(import("@vitest/eslint-plugin")),
            // @ts-expect-error missing types
            interopDefault(import("eslint-plugin-no-only-tests")),
        ] as const);

        pluginTest = pluginTest || {
            ...vitestPlugin,
            rules: {
                ...vitestPlugin.rules,
                // extend `test/no-only-tests` rule
                ...noOnlyTestsPlugin.rules,
            },
        };

        return [
            {
                name: "anolilab/vitest/setup",
                plugins: {
                    vitest: pluginTest,
                },
            },
            {
                files,
                ...tsconfigPath
                    ? {
                        ...vitestPlugin.configs.env,
                        settings: {
                            vitest: {
                                typecheck: true,
                            },
                        },
                    }
                    : {},
                languageOptions: {
                    globals: {
                        ...vitestGlobals,
                    },
                },
                name: "anolilab/vitest/rules",
                rules: {
                    ...vitestPlugin.configs.all.rules,
                    ...vitestPlugin.configs.recommended.rules,

                    "@typescript-eslint/explicit-function-return-type": "off",

                    // Disables
                    "antfu/no-top-level-await": "off",

                    "n/prefer-global/process": "off",

                    "no-unused-expressions": "off",

                    "vitest/consistent-test-it": ["error", { fn: "it", withinDescribe: "it" }],

                    // Enforce a maximum number of expect per test
                    // https://github.com/veritem/eslint-plugin-vitest/blob/main/docs/rules/max-expects.md
                    // This rule should be set on the root config
                    "vitest/max-expects": "off",

                    // Enforce valid expect() usage
                    // https://github.com/veritem/eslint-plugin-vitest/blob/main/docs/rules/no-hooks.md
                    "vitest/no-hooks": "off",

                    // Disallow setup and teardown hooks
                    // https://github.com/veritem/eslint-plugin-vitest/blob/main/docs/rules/no-mocks-import.md
                    "vitest/no-mocks-import": "off",

                    "vitest/no-only-tests": isInEditor ? "off" : "error",

                    // Disallow importing from mocks directory
                    // https://github.com/veritem/eslint-plugin-vitest/blob/main/docs/rules/no-restricted-vi-methods.md
                    "vitest/no-restricted-vi-methods": "off",

                    // Disallow specific vi. methods
                    // https://github.com/veritem/eslint-plugin-vitest/blob/main/docs/rules/no-standalone-expect.md
                    "vitest/no-standalone-expect": "error",

                    // Disallow using expect outside of it or test blocks
                    // https://github.com/veritem/eslint-plugin-vitest/blob/main/docs/rules/valid-expect.md
                    "vitest/valid-expect": ["error", { alwaysAwait: true, maxArgs: 2, minArgs: 1 }],

                    ...overrides,

                    ...prettier
                        ? {
                            "vitest/padding-around-after-all-blocks": "off",
                            "vitest/padding-around-after-each-blocks": "off",
                            "vitest/padding-around-all": "off",
                            "vitest/padding-around-before-all-blocks": "off",
                            "vitest/padding-around-before-each-blocks": "off",
                            "vitest/padding-around-describe-blocks": "off",
                            "vitest/padding-around-expect-blocks": "off",
                            "vitest/padding-around-test-blocks": "off",
                        }
                        : {},
                },
            },
        ];
    },
);
