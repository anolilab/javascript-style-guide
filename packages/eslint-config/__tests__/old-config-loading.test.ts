import { exec } from "node:child_process";
import { rm } from "node:fs/promises";
import { join } from "node:path";
import { promisify } from "node:util";

import { afterEach, beforeEach, describe, expect,it} from "vitest";

const ONE_SECOND_IN_MS = 1000;

const TEST_DIR = join(__dirname, "../__fixtures__", "old-config");
const NODE_MODULES = join(TEST_DIR, "node_modules");

const execAsync = promisify(exec);

describe("integration - old config", () => {
    beforeEach(async () => {
        await rm(NODE_MODULES, { force: true, recursive: true });
    });

    afterEach(async () => {
        await rm(NODE_MODULES, { recursive: true });
    });

    it(
        "installs & works",
        async () => {
            await execAsync("pnpm --ignore-workspace i", { cwd: TEST_DIR });

            const { stderr, stdout } = await execAsync("pnpm exec eslint -c ./.eslintrc.js . || true", {
                cwd: TEST_DIR,
                env: { ...process.env, FORCE_COLOR: "0" },
            });

            expect(stderr.replace(TEST_DIR, "mocked-root-dir")).toMatchSnapshot("stderr", "stderr should be empty");
            expect(stdout.replace(TEST_DIR, "mocked-root-dir")).toMatchSnapshot("stdout", "stdout should show 4 errors");
        },
        240 * ONE_SECOND_IN_MS,
    );
});
