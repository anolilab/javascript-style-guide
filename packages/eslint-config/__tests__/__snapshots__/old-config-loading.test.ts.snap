// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration - old config > installs & works > stderr should be empty 1`] = `""`;

exports[`integration - old config > installs & works > stdout should show 4 errors 1`] = `
"
Your package.json container dependencies for the \\"mdx\\" eslint-plugin, please add the following dependencies with your chosen package manager to enable this plugin:
    eslint-plugin-mdx

To disable this message, add the following to your package.json:
    \\"anolilab\\": { \\"eslint-config\\": { plugin: { \\"plugin-name\\": false } } }

To disable all logging, add the following to your eslint command call \\"NO_LOGS=true eslint ...\\"

mocked-root-dir/.eslintrc.js
  1:24  error  \\"@anolilab/eslint-config\\" is not published  n/no-unpublished-require
  2:25  error  \\"@anolilab/eslint-config\\" is not published  n/no-unpublished-require

/home/prisi/WebstormProjects/anolilab/javascript-style-guide/packages/eslint-config/__fixtures__/old-config/index.js
  1:1  error    Too many blank lines at the beginning of file. Max of 0 allowed  no-multiple-empty-lines
  2:1  warning  Unexpected console statement                                     no-console

âœ– 4 problems (3 errors, 1 warning)
  1 error and 0 warnings potentially fixable with the \`--fix\` option.

"
`;
